### DISCUSSION TITLE
vite-bundling-visual-studio-code-extension/

### DISCUSSION THREAD

**Discussion by giscus** [estruyf/eliostruyf-comments#619] [Comments]
# vite-bundling-visual-studio-code-extension/

Azure / Microsoft 365 / Visual Studio Code / Speaker / Blogger

https://www.eliostruyf.com/vite-bundling-visual-studio-code-extension/

**Comment by Simonl9l** [comment]
tried this with an LSP based extension but call to launch the Language Client in `activate(context: vscode.ExtensionContext) Executable` I get a `[Extension Host] TypeError: Cannot read properties of undefined (reading 'stdio')`

Im assuming there is a bundling issue given the `TransportKind.stdio` (which is an TS emum) as follows in the `vscode-languageserver-protocol/node` package
defied as follows:
```ts
export declare enum TransportKind {
    stdio = 0,
    ipc = 1,
    pipe = 2,
    socket = 3
}
```
my `package.json` has the following dependencies follows:
 "devDependencies": {
    "@types/vscode": "^1.84.0",
    "eslint": "^8.53.0",
    "typescript": "^5.2.2",
    "vite": "^4.5.0",
    "vite-plugin-eslint": "^1.8.1",
    "vite-plugin-static-copy": "^0.17.0"
  },
  "dependencies": {
    "@types/node": "^20.8.10",
    "vscode-languageclient": "^9.0.1",
    "vscode-languageserver-protocol": "^3.17.5",
    "vscode-jsonrpc": "^8.2.0"
  }

```ts

  try {

        let serverModule = context.asAbsolutePath('./out/Server'); 
   
        let debugOptions =  ['--wait-for-debugger'] ; 
        let runExe : Executable = {
            command: serverModule,
            transport: TransportKind.stdio,
            options: {
                env: process.env,
                shell: true,
            }
        };

        let debugExe : Executable = {
            command: serverModule,
            transport: TransportKind.stdio,
            args: debugOptions,
            options: {
                env: process.env,
                shell: true,
            }
        }
        // If the extension is launched in debug mode then the debug server options are used
        // Otherwise the run options are used
        let serverOptions : ServerOptions = {
            run: runExe,
            debug: debugExe
        };

        // Options to control the language client
        let clientOptions: LanguageClientOptions = {
            // Register the server for plain text documents
            documentSelector: [{ scheme: 'file', language: 'plaintext' }],
            synchronize: {
                // Notify the server about file changes to '.csproj files contained in the workspace
                fileEvents: workspace.createFileSystemWatcher('**/.{my ext}')
            }
        };

        // Create the language client and start the client.
        client = new LanguageClient(
            '{my language}',
            '{my description}',
            serverOptions,
            clientOptions
        );

        client.registerProposedFeatures();

        // Start the client. This will also launch the server
        client.start();
    }
    catch(error)
    {
        console.log(error)
    }
```

**Reply by estruyf** [reply]
Thanks, @Simonl9l, for mentioning this. As I wanted to investigate the issue, I started by creating a similar project and installed the latest version of `Vite`, when I noticed that Vite is going to deprecate the `CJS` builds in version 6. In the latest version, they return the warning message: `The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.`

VSCode is not yet ready to make use of `ESM`.

**Reply by ejfasting** [reply]
This will probably change/be fixed when vscode is bumping to electron 28 (?): 
https://github.com/microsoft/vscode/issues/201935 
Electron 28 (finally) support ESM, so we are not longer stuck on commonJS!

It would be interesting with a followup to this article when the time comes! Possibly including an LSP server..?

//ej

**Reply by estruyf** [reply]
Indeed, I'm waiting for this change as well. 🤞

**Reply by estruyf** [reply]
Good news: https://github.com/microsoft/vscode/pull/203956

**Reply by ejfasting** [reply]
*partyhorn*

It would be interesting with an article on how to make use of this change. For instance creating a new extenison, move it from commonJS to ESM and use VITE to bundle it :)

**Reply by estruyf** [reply]
For now, and probably for a longer time, we'll have to stick with CommonJS. I learned last week that the [VSCode loader](https://github.com/microsoft/vscode-loader/) needs to support ESM before we can use it in extensions. Someone already created an issue to track this: https://github.com/microsoft/vscode-loader/issues/56

**Reply by ejfasting** [reply]
ahà, thats a bummer. So this means Vite is out of the question for a while then (?)

**Reply by estruyf** [reply]
I'm afraid the only thing we can do right now is wait. `esbuild` or [tsup](https://tsup.egoist.dev/) are both two good alternatives.

**Comment by tofrankie** [comment]
Hi, I'm trying to build a vscode webview extension using vite. When I debug and run `Run Extension`, `Building` is always displayed below, causing the `Extension Development Host` to fail to open.

Only when one of the tasks is manually turned off will the `Extension Development Host` be turned on.

There was no problem building with webpack before, but after switching to vite, this problem occurred. I suspect that vscode thinks that my task has not been started yet, so it keeps showing `Building`.

`launch.json`

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Run Extension",
      "type": "extensionHost",
      "request": "launch",
      "args": ["--extensionDevelopmentPath=${workspaceFolder}"],
      "outFiles": ["${workspaceFolder}/dist/**/*.js"],
      "preLaunchTask": "run extension",
      "sourceMaps": true
    }
  ]
}
```

`tasks.json`

```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "run ext",
      "type": "shell",
      "command": "pnpm run start:ext",
      "problemMatcher": "$ts-webpack-watch",
      "isBackground": true,
      "presentation": {
        "reveal": "never",
        "group": "watchers"
      },
      "group": {
        "kind": "build"
      }
    },
    {
      "label": "run webview-ui",
      "type": "shell",
      "command": "pnpm run start:webview-ui",
      "problemMatcher": "$ts-webpack-watch",
      "isBackground": true,
      "presentation": {
        "reveal": "never",
        "group": "watchers"
      },
      "group": {
        "kind": "build"
      }
    },
    {
      "label": "run extension",
      "dependsOn": ["run ext", "run webview-ui"]
    }
  ]
}
```

extension `package.json`

```json
{
  "scripts": {
    "package": "vsce package --no-dependencies",
    "prepackage": "run-s build:webview-ui build:ext",
    "start:ext": "vite build --watch",
    "build:ext": "vite build",
    "start:webview-ui": "cd webview-ui && pnpm start",
    "build:webview-ui": "cd webview-ui && pnpm build",
    "lint": "run-s lint:eslint lint:prettier",
    "lint:eslint": "eslint 'src/**/*.{js,jxs}' --cache --fix",
    "lint:prettier": "prettier . '!**/pnpm-lock.yaml' --cache --write",
    "lint:stylelint": "stylelint 'src/**/*.css' --cache --fix"
  }
}
```

webview-ui `package.json`

```json
{
  "scripts": {
    "start": "vite",
    "build": "vite build"
  }
}
```

Do you know the reason? Thanks!

👉 My repo: [github-blogger](https://github.com/toFrankie/github-blogger/tree/dev)

**Reply by estruyf** [reply]
Hi @toFrankie, it's because there is no `problemMatcher` for Vite, so VSCode doesn't know when the watcher completed a new build. I must say, I never use the `preLaunchTask` in the `launch.json`, because that way, I can just start the watch command myself and open/close/refresh without waiting for the watcher to start up.

**Reply by tofrankie** [reply]
Thx
